### PR TITLE
Refactor bootstrapper flags

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

⚠️ Breaking changes ahead!

**What this PR does / why we need it**:
This PR streamlines bootstrapper flags and provides a single point to get and configure a bootstrap cluster.
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NEW: `--bootstrap-type` defaults to `none` and can be extended later to offer more types of bootstrappers.
NEW: `--bootstrap-flags` is a list of flags that are passed to the chosen bootstrapper, if supported.
CHANGE: `--existing-bootstrap-cluster-kubeconfig` is renamed to `--bootstrap-cluster-kubeconfig`
CHANGE: `--bootstrap-cluster-cleanup` replaces `--cleanup-bootstrap-cluster`.
CHANGE: `--minikube` is removed, please use `--bootstrap-flags`.
CHANGE: `--vmdriver` is removed, please use `--bootstrap-flags`.
```

### CODE REVIEW

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (roberthbailey) [inline cmd/clusterctl/clusterdeployer/bootstrap/options.go]: it's 2019 now

File: cmd/clusterctl/cmd/delete_cluster.go [hunk 2]
```diff
@@ -19,13 +19,11 @@ package cmd
 import (
 	"fmt"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (roberthbailey) [inline cmd/clusterctl/cmd/delete_cluster.go]: nit: I don't think this is necessary.

Comment (vincepri) [reply cmd/clusterctl/cmd/delete_cluster.go]: I _think_ this is just goimports being funny :)

Comment (vincepri) [reply cmd/clusterctl/cmd/delete_cluster.go]: Yeah confirmed it's `goimports`, also updated to the latest version to make sure.

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go (hunk 1 above)
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/bootstrap/options.go]: Hehe! Fixed

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+type Options struct {
+	Type             string
+	Cleanup          bool
+	MiniKubeFlags    []string
+	MinikubeVmDriver string
+	KubeConfig       string
+}
+
+func (o *Options) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVarP(&o.Type, "bootstrap-type", "", "minikube", "The cluster bootstrapper to use")
```
Comment (neolit123) [inline cmd/clusterctl/clusterdeployer/bootstrap/options.go]: it was discussed in the meeting today the we shouldn't have a minikube (or even kind) as a default option. possibly the default option should be "none" (in a way).
could we instead break the CLI execution and say something like "hey, you didn't provide a kubeconfig and didn't provide a boostrapper".

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	fs.BoolVarP(&o.Cleanup, "bootstrap-cluster-cleanup", "", true, "Whether to cleanup the bootstrap cluster after bootstrap")
+	fs.StringVarP(&o.KubeConfig, "bootstrap-cluster-kubeconfig", "e", "", "Sets the bootstrap cluster to be an existing Kubernetes cluster.")
+	fs.StringSliceVarP(&o.MiniKubeFlags, "minikube-flags", "", []string{}, "Command line flags to be passed to minikube")
```
Comment (neolit123) [inline cmd/clusterctl/clusterdeployer/bootstrap/options.go]: i guess we can make this the generic `boostrap-flags` and support `kind` using the same field?

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	fs.StringVarP(&o.MinikubeVmDriver, "minikube-vm-driver", "", "", "Adds --vm-driver flag to minikube, for more information refer to https://github.com/kubernetes/minikube/blob/master/docs/drivers.md")
```
Comment (neolit123) [inline cmd/clusterctl/clusterdeployer/bootstrap/options.go]: so this is specific to minikube, can we handle it using the above proposed `boostrap-flags`?
i personally don't see the benefit to "wrap" the `--vm-driver` flag in such a way.

Comment (roberthbailey) [reply cmd/clusterctl/clusterdeployer/bootstrap/options.go]: +1 on not wrapping specific flags to minikube. If there is something we need to change about how we manage minikube I could see a flag for that, but doing a passthrough doesn't seem particularly useful.

Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/bootstrap/options.go]: Fixed!

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go (hunk 4 above)
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/bootstrap/options.go]: Fixed!

File: cmd/clusterctl/clusterdeployer/bootstrap/options.go (hunk 3 above)
Comment (vincepri) [reply cmd/clusterctl/clusterdeployer/bootstrap/options.go]: Fixed, great feedback, thank you!

File: cmd/clusterctl/clusterdeployer/bootstrap/provisioner.go [hunk 6]
```diff
@@ -16,9 +16,28 @@ limitations under the License.
...
+func Get(o Options) (ClusterProvisioner, error) {
+	switch o.Type {
+	case "minikube":
+		return minikube.WithOptions(o.ExtraFlags), nil
+	default:
+		if o.KubeConfig != "" {
+			return existing.NewExistingCluster(o.KubeConfig)
+		}
+
+		return nil, errors.New("Cannot get bootstrapper, specify `--bootstrap-cluster-kubeconfig` to use an existing Kubernetes cluster or specify `--bootstrap-type` to use a built-in ephemeral cluster.")
```
Comment (neolit123) [inline cmd/clusterctl/clusterdeployer/bootstrap/provisioner.go]: errors should start lowercase.
possibly change the error text to `no bootstrapper was specified, alternatively you can ....`?

**Comment by neolit123** [comment]
LGTM on the flags side, thanks @vincepri

**Pull request merged**